src/kudu/consensus/replicate_msg_wrapper.h

```diff
@@ -0,0 +1,193 @@
+#include "kudu/consensus/ref_counted_replicate.h"
+#include "kudu/gutil/strings/substitute.h"
+#include "kudu/util/compression/compression.pb.h"
+#include "kudu/util/compression/compression_codec.h"
+#include "kudu/util/crc.h"
+#include "kudu/util/faststring.h"
+
+namespace kudu {
+namespace consensus {
+
+/**
+ * Thin wrapper to handle compression/decompression of replicate msg
+ */
+class ReplicateMsgWrapper {
+public:
+  ReplicateMsgWrapper(const ReplicateRefPtr &msg,
+                      const CompressionCodec *codec_hint) {
+    orig_msg_ = msg;
+    const CompressionType msg_codec_type =
+        orig_msg_->get()->write_payload().compression_codec();
+    if (msg_codec_type == NO_COMPRESSION) {
+      msg_ = orig_msg_;
+      codec_ = codec_hint;
+    } else {
+      compressed_msg_ = orig_msg_;
+      CHECK_OK(GetCompressionCodec(msg_codec_type, &codec_));
+    }
+    should_compress_ =
+        orig_msg_->get()->op_type() == WRITE_OP_EXT && codec_ != nullptr;
+    DCHECK(msg_ || compressed_msg_);
+  }
+
+  Status Init(faststring *compression_buffer = nullptr) {
+    DCHECK(msg_ || compressed_msg_);
+    if (should_compress_ && msg_ && !compressed_msg_) {
+      return CompressMsg(compression_buffer);
+    }
+    if (compressed_msg_ && !msg_) {
+      return UncompressMsg(compression_buffer);
+    }
+    return Status::OK();
+  }
+
+  ReplicateRefPtr Get() const { return orig_msg_; }
+
+  ReplicateRefPtr Msg() const { return msg_; }
+
+  ReplicateRefPtr CompressedMsg() const { return compressed_msg_; }
+
+private:
+
+  Status UncompressMsg(faststring *buffer = nullptr) {
+    if (!compressed_msg_) {
```

Comment:
is this check, the right check or should it be presence of msg_?

Reply:
We always uncompress the `compressed_msg_`, we take care of populating `compressed_msg_` in the constructor if we know that the payload is compressed (by peeking into the header).